### PR TITLE
CO-1734: adds openapi doc

### PR DESCRIPTION
**Pull request by jemisonf**

### CODE REVIEW

File: openapi.yaml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                  type: string
+                  example: ""
+                expires_in:
+                  description: the number of seconds until this token expires
+                  type: integer
+                refresh_count:
+                  type: integer
+                status:
+                  type: string
+                  enum: [approved]
```
Comment (jemisonf) [inline openapi.yaml]: the descriptions on these properties are kind of my best guess based on looking at the response from apigee; was hoping to get feedback on these, particularly the ones missing descriptions

File: openapi.yaml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+          type: string
+          in: formData
+          description: OAuth Client ID for the application (also called "Consumer Key" in the developer portal)
+        - name: client_secret
+          type: string
+          in: formData
+          description: OAuth Client Secret for the application (also called "Consumer Secret" in the developer portal)
+        - name: grant_type
+          type: string
+          enum: [client_credentials]
```
Comment (jemisonf) [inline openapi.yaml]: not sure if it's possible to have values besides "client_credientials" for this or if we'd want people to use other values

File: openapi.yaml [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+produces:
+  - application/json
+consumes:
+  - application/x-www-form-urlencoded
+paths:
+  /oauth2/token:
+    post:
+      description: Get an access token to make API calls. Part of OAuth2 client credentials flow.
+      consumes:
+        - application/x-www-form-urlencoded
```
Comment (jemisonf) [inline openapi.yaml:25]: I was going to write this to use a JSON body but it looks like it's not possible? I checked the [postman scripts](https://github.com/osu-mist/postman-tools/blob/master/authentication-pre-request.js#L19) as well and it looks they're using `application/x-www-form-urlencoded` too

File: openapi.yaml [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        - name: client_secret
+          type: string
+          in: formData
+          description: OAuth Client Secret for the application (also called "Consumer Secret" in the developer portal)
+        - name: grant_type
+          type: string
+          enum: [client_credentials]
+          in: formData
+      responses:
+        200:
```
Comment (tsoliangwu0130) [inline openapi.yaml]: 400, 401, 500 are also possible error codes.

File: openapi.yaml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+          description: authenticated successfully
+          schema:
+            properties:
+                refresh_token_expires_in:
+                  type: integer
```
Comment (tsoliangwu0130) [inline openapi.yaml]: Looks like it's a string.

File: openapi.yaml [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                  type: string
+                  pattern: .*@.*
+                  example: example@example.com
+                token_type:
+                  description: the type of this token
+                  type: string
+                  enum: [BearerToken]
+                issued_at:
+                  description: the unix timestamp when this token was issued
+                  type: integer
```
Comment (tsoliangwu0130) [inline openapi.yaml]: Should be string as well.

File: openapi.yaml [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                  type: string
+                application_name:
+                  description: the name of the application that requested this token
+                  type: string
+                scope:
+                  type: string
+                  example: ""
+                expires_in:
+                  description: the number of seconds until this token expires
+                  type: integer
```
Comment (tsoliangwu0130) [inline openapi.yaml]: Should be string as well.

File: openapi.yaml [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                refresh_count:
+                  type: integer
```
Comment (tsoliangwu0130) [inline openapi.yaml]: Should be string as well.

**Review by tsoliangwu0130 — CHANGES_REQUESTED** [review]
Looks like this API doest follow the standard of JSON API at all, we need to revisit it at some point.

File: openapi.yaml [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+                    type: string
+                    example: Locations
+                organization_name:
+                  description: the name of the Apigee organization this token was issued by
+                  type: string
+                  enum: [oregonstateuniversity]
+                developer.email:
+                  description: the email of the developer account that requested this token
+                  type: string
+                  pattern: .*@.*
```
Comment (ianbrown9475) [inline openapi.yaml]: Up to you, but you could also use `format: email` here

File: openapi.yaml [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+                  type: string
+                  enum: [BearerToken]
+                issued_at:
+                  description: the unix timestamp when this token was issued
+                  type: string
+                client_id:
+                  description: the client ID of the token requester
+                  type: string
+                access_token:
+                  description: the token issued by the server
```
Comment (ianbrown9475) [inline openapi.yaml]: Could you put something here that tells the user that this field (`access_token`) is the one they need to include in API requests?

File: openapi.yaml [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+          $ref: "#/responses/500"
+responses:
+  '400':
+    description: Incorrect syntax in request, or required fields missing
+    schema:
+      $ref: '#/definitions/ErrorResult'
+  '401':
+    description: User not authorized to perform this action
+    schema:
+      $ref: '#/definitions/ErrorResult'
```
Comment (ianbrown9475) [inline openapi.yaml:107]: Our oauth2 endpoint doesn't respond with JSON:API errors. Instead, they look like this:
```json
{"ErrorCode" : "invalid_request", "Error" :"Required param : grant_type"}
```
or
```json
{"ErrorCode" : "invalid_client", "Error" :"Client credentials are required"}
```
Could you update these schemas?

File: openapi.yaml [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+                  type: string
+                  example: ""
+                expires_in:
+                  description: the number of seconds until this token expires
+                  type: string
+                refresh_count:
+                  type: string
+                status:
+                  type: string
+                  enum: [approved]
```
Comment (ianbrown9475) [inline openapi.yaml]: Is this guaranteed to always be approved? @tsoliangwu0130 might know more about this.

**Review by ianbrown9475 — CHANGES_REQUESTED** [review]

File: openapi.yaml (hunk 10 above)
Comment (jemisonf) [reply openapi.yaml]: that's a good idea 👍

File: openapi.yaml (hunk 12 above)
Comment (jemisonf) [reply openapi.yaml]: was wondering about this as well. I had similar questions for `scope` and `token_type`.

Comment (tsoliangwu0130) [reply openapi.yaml]: Another option will be `revoked`.

File: openapi.yaml (hunk 2 above)
Comment (tsoliangwu0130) [reply openapi.yaml]: Could also be `refresh_token`.

**Review by tsoliangwu0130 — CHANGES_REQUESTED** [review]
More info: https://docs.apigee.com/api-platform/reference/policies/oauthv2-policy

**Review by tsoliangwu0130 — APPROVED** [review]

**Pull request merged**